terraform/layers/app/iam.tf

```diff
@@ -0,0 +1,52 @@
+# data source to assume IAM role policy
+data "aws_iam_policy_document" "ec2_instance_assume_role_policy" {
+  statement {
+    actions = ["sts:AssumeRole"]
+    effect  = "Allow"
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+  }
+}
+
+# Create instance IAM role
+resource "aws_iam_role" "ec2_instance_iam_role" {
+  name               = "${var.project_name}-${var.environment}-ec2-instance-iam-role"
+  assume_role_policy = data.aws_iam_policy_document.ec2_instance_assume_role_policy.json
+}
+
+
+# IAM Policy restricts access to only S3 Bucket hosting web server files
+# IAM Policy restricted to S3 Actions that permit copying files to EC2 only
+resource "aws_iam_role_policy" "s3_to_ec2_policy" {
+  name = "${var.project_name}-${var.environment}-s3-to-ec2-policy"
+  role = aws_iam_role.ec2_instance_iam_role.id
+
+  policy = jsonencode({
+    Version = "2012-10-17",
+    Statement = [
+      {
+        Sid    = "AllowListAndReadS3Objects",
+        Effect = "Allow",
+        Action = [
+          "s3:GetObject",
+          "s3:ListBucket"
+        ],
+        Resource = [
+          "arn:aws:s3:::${var.project_name}-${var.environment}-ec2-website-files",
+          "arn:aws:s3:::${var.project_name}-${var.environment}-ec2-website-files/*"
```

Comment:
for theses 2 arns we can reference the data block directly in the s3.tf file you have. It will be something like 
`   Resource = [
      data.aws_s3_bucket.grad_lab_1_web_files_bucket.arn,
      "${data.aws_s3_bucket.grad_lab_1_web_files_bucket.arn}/*"
    ]`
This way we only have to update the data resource if the name changes rather than all the other references throughout the code